### ISSUE TITLE
Implement full text search

### ISSUE THREAD

**Issue by mariuswilms** [atelierdisko/dsk#44]
The full text search will allow to search over the whole corpus of documents and their properties. We'll be using https://github.com/blevesearch/bleve to drive the search feature. We'll be making some compromises in areas that affect scale as we don't yet have users with large design systems and search result quality as this can be optimized later in an ongoing effort, to get this feature of the door in a timely manner. 

> Search is hard, much harder than you think, [...]

Udi Manber, [Source](https://www.theguardian.com/technology/2007/may/18/news.newmedia)

Pull requests should be submitted against the 1.1 branch of DSK and implemented in APIv2 (already initialized). APIv1 stays frozen. The `search` branch (https://github.com/atelierdisko/dsk/tree/search) has some ideas in form of code comments, where to place functionality (i.e. https://github.com/atelierdisko/dsk/commit/2e0b0e770ea8d86903166e17c0927e110e37b8b0).

The entry point for this will be a new HTTP API `search` endpoint, the previous `search` endpoint will be renamed to `filter`. (Happened in https://github.com/atelierdisko/dsk/commit/a610bc2ff35c53ef7b0ff6d321a6bac12f483e79#diff-9a8b58ee2383bb1bb323bca70e2eba65R52)

Queries are provided in using the `q` HTTP query argument: `/search?q={query}`, the argument must be escaped in a way that it unescapes correctly using https://golang.org/pkg/net/url/#QueryUnescape. Search queries will support multiple words and use `Match` queries (http://blevesearch.com/docs/Query/). We'll support what bleve supports out of the box, nothing fancy, no facets.

As a simplification the index will be an in-memory index, this way we don't need to deal with additional files and storage locations. This also means the index must be created each time DSK starts up. We expect ~10s for creating a medium sized design system. The initial indexing will not block the startup but executed asynchronously.

In general the frontend and backend should be able to deal with stale search results, (entirely) missing results or results for nodes that don't exist anymore. 

With the addition of the search index we'll have to keep four states in sync with each other. Keeping the frontend in sync is not a strict requirement, but an additional feature supported by the builtin frontend.

1. the documents on disk
2. the structures build from documents on disk (aka `Node` and `NodeTree`)
3. the search index
4. the frontend (i.e. via WebSockets)

The search index should be rebuild after a change was detected similar to how this is done in the `NodeTree` (https://github.com/atelierdisko/dsk/blob/1.1/node_tree.go#L178-L180). We should think about waiting for the tree rebuild to be settled and then start indexing.

We'll ignore language of the documents for now as design system documents generally use a mixture of english and other languages. For this to be detected correctly the bleve library would need to be build with a C extension, thus complicate the build process of DSK.

**Comment by mariuswilms** [comment]
The feature has been implemented in https://github.com/atelierdisko/dsk/commit/c60e3b536e8b7cf837e928268658366884de9af0 with support by @wegry

I'll leave this issue open until we deem the feature ready to be released with 1.1, 1.1 release on the other hand is blocked until this feature is ready.

The following tasks are open:

- [ ] Stress test the implementation, file bugs and add more tests
- [ ] It should be possible to search for multiple words
- [ ] The DSK client needs to be updated to use API v2
- [ ] It should be possible to search for the title
- [ ] ...the filenames of downloads

**Comment by mariuswilms** [comment]
I took a different path in regards to the fuzzy flag. All matches should always be perforemed with a bit of fuzzyness to cater for typos. Prefix search should be limited to filtering only. 

I was able to check all the boxes, with the exception of language support. This is tricky to implement with bleve https://groups.google.com/d/msg/bleve/W05zDao3-gI/-r4pOzmlAwAJ might help in the future. 

Currently the search query input will be analyzed using the primary language only, that is the language that is provided first using the `-lang` command line option. This is a limitation I can live with.
